src/library/scala/collection/mutable/HashTable.scala

```diff
@@ -357,14 +360,14 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
 
   protected def elemEquals(key1: A, key2: A): Boolean = (key1 == key2)
 
-  // Note:
-  // we take the most significant bits of the hashcode, not the lower ones
-  // this is of crucial importance when populating the table in parallel
-  protected final def index(hcode: Int) = {
+  /**
+    * Note: we take the most significant bits of the hashcode, not the lower ones
+    * this is of crucial importance when populating the table in parallel
+    */
+  protected final def index(hcode: Int): Int = {
     val ones = table.length - 1
-    val improved = improve(hcode, seedvalue)
-    val shifted = (improved >> (32 - java.lang.Integer.bitCount(ones))) & ones
-    shifted
+    val exponent = Integer.numberOfLeadingZeros(ones)
+    (improve(hcode, seedvalue) >>> exponent) & ones
```

Comment:
Oh I'm not paying enough attention tonight, it seems. This is the `index` method, which *needs* to return `0` for correctness when `length == 1` (assuming that ever happens).

I don't know whether it is reasonable to assume that `length` is never 1. If it is, then we can drop `& ones`. If it isn't, we need to keep the code as it currently is, but then it would probably be a good thing to add a comment that `& ones` is there specifically to handle `length == 1`, in case someone wonders in the future.